c7n/resources/account.py

```diff
@@ -2497,3 +2497,96 @@ def process(self, resources):
             client.put_model_invocation_logging_configuration(loggingConfig=params)
         else:
             client.delete_model_invocation_logging_configuration()
+
+
+@filters.register('ec2-metadata-defaults')
+class EC2MetadataDefaults(ValueFilter):
+    """Filter on the default instance metadata service (IMDS) settings for the specified account and
+    region.  NOTE: Any configuration that has never been set (or is set to 'No Preference'), will
+    not be returned in the response.
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: ec2-imds-defaults
+                resource: account
+                filters:
+                - or:
+                  - type: ec2-metadata-defaults
+                    key: HttpTokens
+                    value: optional
+                  - type: ec2-metadata-defaults
+                    key: HttpTokens
+                    value: absent
+    """
+
+    annotation_key = 'c7n:EC2MetadataDefaults'
+    annotate = False  # no annotation from value filter
+    schema = type_schema('ec2-metadata-defaults', rinherit=ValueFilter.schema)
+    permissions = ('ec2:GetInstanceMetadataDefaults',)
+
+    def augment(self, resources):
+        client = local_session(self.manager.session_factory).client('ec2')
+        for r in resources:
+            r[self.annotation_key] = self.manager.retry(
+                client.get_instance_metadata_defaults)["AccountLevel"]
+
+    def process(self, resources, event=None):
+        self.augment([r for r in resources if self.annotation_key not in r])
+        return super(EC2MetadataDefaults, self).process(resources, event)
+
+    def __call__(self, r):
+        return super(EC2MetadataDefaults, self).__call__(r[self.annotation_key])
+
+
+@actions.register('set-ec2-metadata-defaults')
+class SetEC2MetadataDefaults(BaseAction):
+    """Modifies the default instance metadata service (IMDS) settings at the account level.
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: set-ec2-metadata-defaults
+                resource: account
+                filters:
+                  - or:
+                    - type: ec2-metadata-defaults
+                      key: HttpTokens
+                      op: eq
+                      value: optional
+                    - type: ec2-metadata-defaults
+                      key: HttpTokens
+                      value: absent
+                actions:
+                  - type: set-ec2-metadata-defaults
+                    HttpTokens: required
+
+    """
+
+    schema = type_schema(
+        'set-ec2-metadata-defaults',
+        HttpTokens={'enum': ['optional', 'required', 'no-preference']},
+        HttpPutResponseHopLimit={'type': 'integer'},
+        HttpEndpoint={'enum': ['enabled', 'disabled', 'no-preference']},
+        InstanceMetadataTags={'enum': ['enabled', 'disabled', 'no-preference']},
+    )
```

Comment:
There's a tension between hardcoding these property names/types/options here and trying to handle things at runtime with shape validation. As written this looks good today, but if we get a new parameter upstream policies won't be able to use it because this schema will cause the policy to fail a structure check before we even get to the shape validation below.

Other operations (like [the `update-trail` action](https://github.com/cloud-custodian/cloud-custodian/blob/41b6496cddba5fedf5f068eb435b85d2e849582c/c7n/resources/cloudtrail.py#L181-L197)) we leave the action schema deliberately loose. That's more flexible but makes `custodian schema` output less useful.

I've tried out a middle-ground option in https://github.com/mattheidelbaugh/cloud-custodian/pull/2 that tries to expand a class's schema property based on the shape data. In theory that makes things a bit more discoverable while _also_ having an operation's schema reflect the latest service model data.

...but it may not be a great idea, or the implementation could be flawed in some way I'm not seeing :sweat_smile: . Curious if you see any issues with the approach or have other ideas.

Reply:
while I agree there is a tension, its more about relevant for things that reasonable will have expansion or large in size. that's not always obvious in advance, but it does seem reasonable in this case that this set of attributes stays relevatively static due to the large surface area of interactions from any api utilizing program. as such, I'm good with this one as is.

Reply:
I love this idea.  I was wrestling with the same problem when I was implementing this.  The only issues I can see so far are:
- handling of the `attributes` key that we sometimes use to encapsulate all of the API params in the policy action
- dropping the resource id key from schema if it's present
- it looks like the some of the type names are slightly different i.e. 'object' is 'structure', 'array' is 'list'